New toolkit structure

anmacdonald opened the pull request:
Initial PR. This covers:

* CoordinateConversion & NorthArrow done for a QML, Quick, and Widgets workflow.
* A structure for importing that requires only the inclusion of a PRI file. Can confirm works on iOS, Android and Desktop platforms.
* New documentation style (doxygen-like).

Needs:
* A better README and more documentation on how-to-use.
* Examples

lsmallwood commented on README.md:
````diff
@@ -1,93 +1,110 @@
-...
-NorthArrow {
+```qml
+import Esri.ArcGISRuntime 100.X
````
The toolkit will be targeting a specific version of Runtime (I think) so we could make this 100.8

lsmallwood commented on README.md:
```diff
@@ -1,93 +1,110 @@
 
 - add an import statement for the toolkit:
-  `import esri.arcgisruntime.toolkit 1.0`
+  `import "qrc:///Esri/ArcGISRuntime/Toolkit" as Toolkit`
```
I wonder if it might be better to promote a name which includes Esri (and to avoid clashes with anything else called toolkit)

 `import "qrc:///Esri/ArcGISRuntime/Toolkit" as ArcGISRuntimeToolkit`

lsmallwood commented on README.md:
```diff
@@ -1,93 +1,110 @@
+      BasemapTopographic {}
+    }
+
+    Toolkit.NorthArrow {
```
Add some comments here to explain what is happening

lsmallwood commented on README.md:
```diff
@@ -1,93 +1,110 @@
 get added to QtCreator when you install the ArcGIS Runime SDK for Qt.
 
-- In QtCreator choose `File/New file or project/ArcGIS/ArcGIS Runtime 100.6 Qt
+- In QtCreator choose `File/New file or project/ArcGIS/ArcGIS Runtime 100.X Qt
```
100.8

lsmallwood commented on README.md:
````diff
@@ -1,93 +1,110 @@
+- in `main.cpp` (inside the main function) call the registration function.
+
+```cpp
+Esri::ArcGISRuntime::Toolkit::registerComponents(myQmlEngine);
````
It might be good to show an example of where `myQmlEngine` comes from

lsmallwood commented on README.md:
````diff
@@ -96,57 +113,60 @@ in your own QML files.
-...
-NorthArrow {
+```qml
+import Esri.ArcGISRuntime 100.X
````
same comments as QML

lsmallwood commented on README.md:
```diff
@@ -96,57 +113,60 @@ in your own QML files.
 
 - in your widgets code, create a new instance of the tool you wish to use and
-  add it your widets layout. You will also normally need to pass the `GeoView`
+  add it your widgets layout. You will also normally need to pass the `GeoView`
```
```suggestion
  add it to your widgets layout. You will also normally need to pass the `GeoView`
```

lsmallwood commented on README.md:
```diff
@@ -96,57 +113,60 @@ in your own QML files.
   ...
   auto northArrow = new NorthArrow(this);
-  northArrow->setGeometry(0,0,48,48); // set size and position for the tool
   northArrow->setMapView(m_mapView);
```
is this not `setGeoView`?

anmacdonald replied:
In the widgets I elected for two: `setMapView` and `setSceneView` functions so there was only ever one _type_ of View that was allowed in those cases. Not sure if good or bad, but in the widget case `setMapView` is correct.

lsmallwood commented on common.pri:
```diff
@@ -0,0 +1,37 @@
+           $$CPPPATH/GenericListModel.cpp \
+           $$CPPPATH/GenericTableProxyModel.cpp \
+           $$CPPPATH/MetaElement.cpp \
+           $$CPPPATH/NorthArrowController.cpp
```
blank line at end of file

lsmallwood commented on cpp/Esri/ArcGISRuntime/Toolkit/CoordinateConversionConstants.cpp:
```diff
@@ -0,0 +1,39 @@
+const QString DEGREES_MINUTES_SECONDS_FORMAT = QStringLiteral("DMS");
+const QString MGRS_FORMAT = QStringLiteral("MGRS");
+const QString USNG_FORMAT = QStringLiteral("USNG");
+const QString UTM_FORMAT = QStringLiteral("UTM");
```
alphabetic

lsmallwood commented on cpp/Esri/ArcGISRuntime/Toolkit/CoordinateConversionConstants.h:
```diff
@@ -0,0 +1,51 @@
+extern const QString DECIMAL_DEGREES_FORMAT;
+extern const QString DEGREES_DECIMAL_MINUTES_FORMAT;
+extern const QString DEGREES_MINUTES_SECONDS_FORMAT;
+extern const QString MGRS_FORMAT;
```
alphabetic

lsmallwood commented on cpp/Esri/ArcGISRuntime/Toolkit/CoordinateConversionController.cpp:
```diff
@@ -0,0 +1,272 @@
+ ******************************************************************************/
+#include "CoordinateConversionController.h"
+
+#include "CoordinateConversionResult.h"
```
we should probably have a comment to describe toolkit headers

lsmallwood commented on cpp/Esri/ArcGISRuntime/Toolkit/CoordinateConversionController.cpp:
```diff
@@ -0,0 +1,272 @@
+
+namespace
+{
+constexpr double DEFAULT_ZOOMTO_DISTANCE = 1500.0;
```
```suggestion
constexpr double DEFAULT_ZOOM_TO_DISTANCE = 1500.0;
```

lsmallwood commented on cpp/Esri/ArcGISRuntime/Toolkit/CoordinateConversionController.cpp:
```diff
@@ -0,0 +1,272 @@
+        {
+          auto index = m_conversionResults->index(i);
+          auto result = m_conversionResults->element<CoordinateConversionResult>(index);
+          if (result)
```
I think you should include the {} for readability in this if/else

lsmallwood commented on cpp/Esri/ArcGISRuntime/Toolkit/CoordinateConversionController.cpp:
```diff
@@ -0,0 +1,272 @@
+  }
+}
+
+CoordinateConversionController::~CoordinateConversionController()
```
```suggestion
CoordinateConversionController::~CoordinateConversionController() = default
```

lsmallwood commented on cpp/Esri/ArcGISRuntime/Toolkit/CoordinateConversionController.cpp:
```diff
@@ -0,0 +1,272 @@
+    {
+      if (m_inPickingMode)
+      {
+        setCurrentPoint(sceneView->screenToBaseSurface(event.x(), event.y()));
```
this won't take account of buildings etc. for the scene view. Probably better to use this than the async call though...

anmacdonald replied:
Agreed. 
This was lifted from the old tool -  if you think this should be changed then this would be the time.

lsmallwood commented on cpp/Esri/ArcGISRuntime/Toolkit/CoordinateConversionController.cpp:
```diff
@@ -0,0 +1,272 @@
+  const QString& p,
+  CoordinateConversionOption* option)
+{
+  if (auto geoView = qobject_cast<GeoView*>(m_geoView))
```
when would this be false?

anmacdonald replied:
`m_geoView` is technically an arbitrary QObject.

lsmallwood commented on cpp/Esri/ArcGISRuntime/Toolkit/CoordinateConversionController.cpp:
```diff
@@ -0,0 +1,272 @@
+  }
+  else if (auto mapView = qobject_cast<MapView*>(m_geoView))
+  {
+    res = mapView->locationToScreen(m_currentPoint);
```
what happens for mapView if it's off the screen?

lsmallwood commented on cpp/Esri/ArcGISRuntime/Toolkit/CoordinateConversionController.cpp:
```diff
@@ -0,0 +1,272 @@
+  auto result = new CoordinateConversionResult(m_conversionResults);
+  result->setType(option);
+  result->updateCoordinatePoint(currentPoint());
+  m_conversionResults->append(result);
```
seems like the append function could be replaced with one which actually does the creation (e.g. take in the type and point)

anmacdonald replied:
This is an interesting idea! A bit like `std::make_shared` and how it takes arbitrary arguments?

lsmallwood commented on cpp/Esri/ArcGISRuntime/Toolkit/CoordinateConversionController.cpp:
```diff
@@ -0,0 +1,272 @@
+  emit currentPointChanged(QVariant::fromValue(static_cast<Geometry>(m_currentPoint)));
+}
+
+
```
remove extra line

lsmallwood commented on cpp/Esri/ArcGISRuntime/Toolkit/CoordinateConversionController.h:
```diff
@@ -0,0 +1,309 @@
+  Q_INVOKABLE QPointF screenCoordinate() const;
+
+  /*!
+   * \brief Distance between camera and current point when zoomToCurrentPoint
```
in meters (?)

anmacdonald replied:
I have commented further down that the units are those defined as distance in the geoView's SpatialReference.

lsmallwood commented on cpp/Esri/ArcGISRuntime/Toolkit/CoordinateConversionController.h:
```diff
@@ -0,0 +1,309 @@
+   * signals. To grab the point returned by this signal, use a
+   * qvariant_cast<Geometry> followed by a geometry_cast<Point>.
+   */
+  void currentPointChanged(QVariant point);
```
should this be const &?

anmacdonald replied:
It's optional. As passed into the QML layer it will always be copied anway. But no harm in cont-ref'ing it!

lsmallwood commented on cpp/Esri/ArcGISRuntime/Toolkit/CoordinateConversionController.h:
```diff
@@ -0,0 +1,309 @@
+  /*!
+   * \brief Emitted when the currentPoint property has changed.
+   * \note This signal is a QVariant containing the Point cast to a Geometry
+   * type. Due to technical limitations a `Point` is not a Qt MetaType
```
as discussed, seems like we could register this in the API though?

anmacdonald replied:
I would like to, all the code saying we do this because of API reasons can be removed by then.

lsmallwood commented on cpp/Esri/ArcGISRuntime/Toolkit/CoordinateConversionController.h:
```diff
@@ -0,0 +1,309 @@
+
+public slots:
+  /*!
+   * \brief Set the current point to point. This updates all textual
```
if we are doing qdoc I guess this should be `to \a point`

anmacdonald replied:
This is also doxygen style so happy with this!

lsmallwood commented on cpp/Esri/ArcGISRuntime/Toolkit/CoordinateConversionController.h:
```diff
@@ -0,0 +1,309 @@
+   * \brief Removes a given CoordinateConversionResult at index.
+   * 
+   * This function is for deleting results in the list-model returned by
+   * coordinateREsults.
```
```suggestion
   * coordinateResults.
```

lsmallwood commented on cpp/Esri/ArcGISRuntime/Toolkit/CoordinateConversionController.h:
```diff
@@ -0,0 +1,309 @@
+private:
+  Point m_currentPoint;
+  double m_zoomToDistance;
+  GenericListModel* m_coordinateFormats;
```
```suggestion
  GenericListModel* m_coordinateFormats = nullptr;
```

for all ptr types

anmacdonald replied:
Defined in constructor.

lsmallwood commented on cpp/Esri/ArcGISRuntime/Toolkit/CoordinateConversionController.h:
```diff
@@ -0,0 +1,309 @@
+  GenericListModel* m_coordinateFormats;
+  GenericListModel* m_conversionResults;
+  QObject* m_geoView;
+  bool m_inPickingMode;
```
```suggestion
  bool m_inPickingMode = false;
```
?

anmacdonald replied:
Defined in constructor.

lsmallwood commented on cpp/Esri/ArcGISRuntime/Toolkit/CoordinateConversionOption.cpp:
```diff
@@ -0,0 +1,257 @@
+  emit nameChanged();
+}
+
+bool CoordinateConversionOption::addSpaces() const
```
I guess this should be `isAddSpaces` - currently it sounds like it will actually perform an add

lsmallwood commented on cpp/Esri/ArcGISRuntime/Toolkit/CoordinateConversionOption.h:
```diff
@@ -0,0 +1,247 @@
+   * This conversion has the chance of failing, in which case an invalid
+   * default-constructed Point will be returned.
+   * 
+   * \param point QString to conver to a Point.
```
```suggestion
   * \param point QString to convert to a Point.
```

lsmallwood commented on cpp/Esri/ArcGISRuntime/Toolkit/CoordinateConversionOption.h:
```diff
@@ -0,0 +1,247 @@
+   * \brief Set the precision.
+   * \sa precision
+   */
+  void setPrecision(int precision);
```
i wonder if decimal places and precision should both be unsigned

anmacdonald replied:
This is a good point, but the `CoordinateFormatter` accepts only ints, so will stick to that.

lsmallwood commented on cpp/Esri/ArcGISRuntime/Toolkit/GenericListModel.cpp:
```diff
@@ -0,0 +1,380 @@
+  return m_elementType;
+}
+
+void GenericListModel::setDisplayPropertyName(QString propertyName)
```
```suggestion
void GenericListModel::setDisplayPropertyName(const QString& propertyName)
```

lsmallwood commented on cpp/Esri/ArcGISRuntime/Toolkit/GeoViews.h:
```diff
@@ -0,0 +1,58 @@
+* but they implement these interfaces independently due to Qt limitations with
+* multiple inheritance.
+*
+* To compound this issue: when the toolkit supports the Quick workflow we
```
```suggestion
* In addition, when the toolkit supports the Quick workflow we
```

lsmallwood commented on cpp/Esri/ArcGISRuntime/Toolkit/GeoViews.h:
```diff
@@ -0,0 +1,58 @@
+* This include will include the correct set of Views depending on 
+* whether flag \c WIDGETS_ARCGISRUNTIME_TOOLKIT is defined or not.
+*
+* Implicitly, this means we assume the user will not need a combination of
```
i don't think we need doc for this part

anmacdonald replied:
What's the objection to this part of the docs?

anmacdonald commented on cpp/Esri/ArcGISRuntime/Toolkit/GenericTableProxyModel.cpp:
```diff
@@ -0,0 +1,210 @@
+  QStringList parts = string.split(' ', QString::SkipEmptyParts);
+  for (auto& part: parts)
+  {
+    part.replace(0, 1, part.at(0).toUpper());
```
None I can find.

anmacdonald commented on cpp/Esri/ArcGISRuntime/Toolkit/CoordinateConversionOption.h:
```diff
@@ -0,0 +1,247 @@
+
+private:
+  QString m_name;
+  CoordinateType m_outputMode;
```
done in constructor.

JamesMBallard replied:
I will double down on these comments. I think it is clearer to have them here and remove them from the constructor.

That being said, if this is existing code that didn't write, I don't want to inundate you with requests for changes.

anmacdonald commented on cpp/Esri/ArcGISRuntime/Toolkit/GenericListModel.cpp:
```diff
@@ -0,0 +1,380 @@
+  else if (!index.isValid())
+    return QVariant();
+
+  auto o = m_objects.at(index.row());
```
This is technically done by the `index::isValid` check. Indices can only be created via the `index` member function of the ListModel, which does bounds checking.

JamesMBallard commented:
@anmacdonald can you provide a list of which existing files have had changes other than structural changes to support the new workflow?

I don't want to review files that haven't changed in any meaningful way and I don't want you to have to defend or justify code that was written by others and you are simply reshuffling it.

JamesMBallard commented on cpp/Esri/ArcGISRuntime/Toolkit/CoordinateConversionController.h:
```diff
@@ -0,0 +1,311 @@
+  GenericListModel* m_coordinateFormats;
+  GenericListModel* m_conversionResults;
+  QObject* m_geoView;
+  bool m_inPickingMode;
```
I would second all these recommendations to define these values in the header.

anmacdonald replied:
2v1 here! I'll change.

JamesMBallard commented on cpp/Esri/ArcGISRuntime/Toolkit/CoordinateConversionOption.h:
```diff
@@ -0,0 +1,248 @@
+class CoordinateConversionOption : public QObject
+{
+  Q_OBJECT
+  Q_CLASSINFO("RegisterEnumClassesUnscoped", "false")
```
What does this mean/do?

anmacdonald replied:
It means in QML Code that enums we declare will have syntax `CoordianteType.Gars` as opposed to defaulting to just `Gars`.

JamesMBallard replied:
That's really nice. Do we use this anywhere else?

JamesMBallard commented on cpp/Esri/ArcGISRuntime/Toolkit/CoordinateConversionResult.h:
```diff
@@ -0,0 +1,139 @@
+   * \sa CoordinateConversionController::currentPointChanged(QVariant point)
+   * \param point point object to store in notation field.
+   */
+  void updateCoordinatePoint(const QVariant& point);
```
See related comments. I think you can add a `Q_DECLARE_METATYPE(Esri::ArcGISRuntime::Point)` to something to avoid this.

JamesMBallard commented on cpp/Esri/ArcGISRuntime/Toolkit/GenericListModel.cpp:
```diff
@@ -0,0 +1,379 @@
+    output.insert(i - offset + Qt::UserRole + 1, property.name());
+  }
+  return output;
+}
```
There's a lot of meta object magic going on here. Can you give a quick overview of how this works and what it's doing?

anmacdonald replied:
In the code? Yes good idea I'll add a comment.

JamesMBallard commented on cpp/Esri/ArcGISRuntime/Toolkit/GenericListModel.cpp:
```diff
@@ -0,0 +1,379 @@
+    if (o)
+      disconnect(o, &QObject::destroyed, this, nullptr);
+
+    delete o;
```
Should we be storing raw pointers if we need to do this? This seems like a red flag.

anmacdonald replied:
We listen into `&QObject::destroyed` signals in the `connectElement` method, so by the time we reach here the pointer will be null if already deleted outside of this model.

What are you suggesting? A list of `QPoint<QObject>` instead? That might actually be quite a bit cleaner than the current system.

JamesMBallard replied:
My comment on this is that it makes me nervous to see raw delete calls. I realize shared pointers probably won't work, but I wonder if there is any other solution to avoid this?

JamesMBallard commented on cpp/Esri/ArcGISRuntime/Toolkit/GenericListModel.cpp:
```diff
@@ -0,0 +1,379 @@
+  {
+    auto property = m_elementType->property(i);
+    auto notifySignal = property.notifySignal();
+    if (notifySignal.isValid())
```
Would https://doc.qt.io/qt-5/qmetaproperty.html#hasNotifySignal be clearer?

anmacdonald replied:
Good find! Yes it would be!

JamesMBallard commented on cpp/Esri/ArcGISRuntime/Toolkit/GenericListModel.cpp:
```diff
@@ -0,0 +1,379 @@
+        index,
+        i - offset + Qt::UserRole +1,
+        object,
+        this);
```
Would there be a lot of these created? I assume with `this` as the parent they won't leak, but there is no mechanism to clean them up early, either.

anmacdonald replied:
In the MetaElement constructor we have the following:

```cpp
  connect(this, &MetaElement::propertyChanged, this, &MetaElement::emitDataChanged);
  connect(m_trackedObject.data(), &QObject::destroyed, this, &QObject::deleteLater);
```

So they will delete themselves if either the model or the desired element in the model also dies.

JamesMBallard commented on cpp/Esri/ArcGISRuntime/Toolkit/GenericListModel.h:
```diff
@@ -0,0 +1,343 @@
+{
+
+/*!
+ * \internal
```
This is a tremendous amount of documentation for an internal class.

anmacdonald replied:
It sure is. In our current workflow it's all part of the user source code and accessible/usable to them. 

While we can inform the user that something is an implementation detail, we can't prevent them from using/inspecting it. I have gone for a blanket approach of documenting everything as if it were fair game to the consumer.

JamesMBallard replied:
@anmacdonald that's fair, but for the classes that are purely implementation details, I would make it obvious other than just `\internal` that these are internal classes and nobody need understand how/why they work or exist.

This level of documentation would grab my eye as something that is paramount that I would need to know about and understand as a user.

JamesMBallard commented on cpp/Esri/ArcGISRuntime/Toolkit/GenericListModel.h:
```diff
@@ -0,0 +1,343 @@
+   * expose.
+   * \param parent Owning parent QObject.
+   */
+  GenericListModel(const QMetaObject* elementType, QObject* parent = nullptr);
```
explicit

JamesMBallard commented on cpp/Esri/ArcGISRuntime/Toolkit/GenericListModel.h:
```diff
@@ -0,0 +1,343 @@
+   * \brief Constructor
+   * \param parent Owning parent QObject.
+   */
+  Q_INVOKABLE GenericListModel(QObject* parent = nullptr);
```
explicit

lsmallwood commented on cpp/Esri/ArcGISRuntime/Toolkit/MetaElement.cpp:
```diff
@@ -0,0 +1,50 @@
+  if (!m_trackedObject || !m_index.isValid())
+    return;
+  else
+    emit m_parentModel->dataChanged(m_index, m_index, { Qt::UserRole, m_customRole });
```
this causes the parent to emit it's own dataChanged signal?

anmacdonald replied:
It does yes!

lsmallwood commented on cpp/Esri/ArcGISRuntime/Toolkit/NorthArrowController.h:
```diff
@@ -0,0 +1,98 @@
+ * 
+ * This controller calculates the current heading from a GeoView, and allows
+ * the NorthArrow to apply a given heading to the GeoView.
+ * 
```
empty line

lsmallwood commented on cpp/Esri/ArcGISRuntime/Toolkit/NorthArrowController.h:
```diff
@@ -0,0 +1,98 @@
+   * Internally this is cast to a MapView or SceneView using `qobject_cast`,
+   * which is why the paremeter is of form `QObject` and not `GeoView`.
+   * 
+   * \param geoView Object which must inherit from GeoView* and QObject*. 
```
wrong param name?

lsmallwood commented on cpp/Esri/ArcGISRuntime/Toolkit/NorthArrowController.h:
```diff
@@ -0,0 +1,98 @@
+
+  /*!
+   * \brief The calculated heading of this controller.
+   * \return heading as double
```
in degrees?

lsmallwood commented on cpp/Esri/ArcGISRuntime/Toolkit/NorthArrowController.h:
```diff
@@ -0,0 +1,98 @@
+public slots:
+  /*!
+   * \brief Set the heading by rotating the MapView or SceneView camera to the
+   * given heading.
```
in degrees

lsmallwood commented on import/CoordinateConversion.qml:
```diff
@@ -0,0 +1,327 @@
+ * \qmltype CoordinateConversion
+ * \inqmlmodule Esri.ArcGISRuntime.Toolkit
+ * \since Esri.ArcGISRutime 100.8
+ * \brief The user interface for the coordinate conversion tool.
```
Should we have a note that this file is shared for qml and cpp? Maybe that just confuses matters

anmacdonald replied:
This is a good point. I was thinking that if we had a `CONTRIBUTING.md` file - we could go into more details on structure across the system there.

anmacdonald commented on README.md:
```diff
@@ -1,152 +1,172 @@
 
 - add an import statement for the toolkit:
-  `import esri.arcgisruntime.toolkit 1.0`
+  `import "qrc:///Esri/ArcGISRuntime/Toolkit" as ArcGISRuntimeToolkit`
```
That'll be the case, I am afraid.

There is the small bonus that this does remove the need for `QML2_IMPORT_PATH`.

I'm not sure there is another way to import statically linked QML that are properly namespaced without moving to Plugins -- which are built separately. I'm willing to move to plugins, but the original requirement was that the user was building this as source code in their current project.

JamesMBallard replied:
I don't think this is too bad. It's not visually appealing, but it's better since we can remove the `QML2_IMPORT_PATH` requirement.

anmacdonald commented on common.pri:
```diff
@@ -0,0 +1,37 @@
+           $$CPPPATH/GenericListModel.cpp \
+           $$CPPPATH/GenericTableProxyModel.cpp \
+           $$CPPPATH/MetaElement.cpp \
+           $$CPPPATH/NorthArrowController.cpp
```
My arm can be twisted on this one, but I'd rather list files explicitly as this is considered good practice.

JamesMBallard replied:
Fair enough, it's not too onerous now so it can remain like this.

anmacdonald commented on cpp/Esri/ArcGISRuntime/Toolkit/CoordinateConversionOption.h:
```diff
@@ -0,0 +1,248 @@
+   * \return String representation of point with the current formatting options
+   * applied.
+   */
+  Q_INVOKABLE QString prettyPrint(const Point& point) const;
```
Not entirely new, it's just been moved from `CoordinateConversionController`. It's called from `CoordinateConversionResult`. It does not need to be `Q_INVOKABLE` so I shall remove that.

anmacdonald commented on cpp/Esri/ArcGISRuntime/Toolkit/CoordinateConversionController.h:
```diff
@@ -0,0 +1,311 @@
+   * \param spatialReference An explicit SpatialReference for conversion.
+   * \param option Option dictating hints on how to convert the string to a
+   *  point.
+   */
```
My opinion differs on that one.

anmacdonald commented on cpp/Esri/ArcGISRuntime/Toolkit/CoordinateConversionOption.h:
```diff
@@ -0,0 +1,248 @@
+   * \return QString as a Point. 
+   */
+  Q_INVOKABLE Point pointFromString(const QString& point,
+                                    const SpatialReference& spatialReference);
```
Removing `Q_INVOKABLE`.

anmacdonald commented on cpp/Esri/ArcGISRuntime/Toolkit/GenericListModel.cpp:
```diff
@@ -0,0 +1,379 @@
+    if (!m_elementType)
+      return;
+
+    if (roles.contains(Qt::DisplayRole))
```
All Qt "roles" are passed and declared as ints to address the user needing to add their own.

anmacdonald commented on cpp/Esri/ArcGISRuntime/Toolkit/GenericListModel.cpp:
```diff
@@ -0,0 +1,379 @@
+       const QVector<int>& roles)
+  {
+    if (m_displayPropIndex < 0)
+      return;
```
This code basically covers: "do we want to throw an additional signal out there on top of the one I am currently receiving?"

It won't suppress the old signal, but we won't send out the additional one.

anmacdonald commented:
@JamesMBallard 

> There must be a way to do this in the tools so we don't need to in the API. That's the whole idea is that folks that need that can do it. Can you add a single "ValueTypeRegistrations.h" or something like that that only adds all the Q_DECLARE_METATYPE(Esri::ArcGISRuntime::Point) and any others you need?
> 
> That should alleviate this.

So I have investigated and I can get it running with:

> Q_DECLARE_METATYPE(Esri::ArcGISRuntime::Point)
and
>   qRegisterMetaType<Point>("Point");

In tandem.

The `qRegisterMetaType` call can go into the `register` function of the toolkit in lieu of somewhere better. But I'd really gun for an API change here vs. giving the following implicit workflow:
* If you want to use `Esri::ArcGISRuntime::Point` in Qt's metatype system, register it yourself.
* Unless you are using toolkit, then expose `Point.h` in your code only through the `"ValueTypeRegistrations.h"` header file.

lsmallwood commented:
> But I'd really gun for an API change here vs. giving the following implicit workflow

I'd agree. We should do this in the API I think

JamesMBallard commented:
@anmacdonald @lsmallwood good points, I agree. Let's get the Point registration into the Qt Api. Point only, and clean up this code once that's done. We will avoid the slippery slope argument and get this done to clean it up.

@anmacdonald can you take that on?

JamesMBallard commented on widgets/Esri/ArcGISRuntime/Toolkit/Flash.cpp:
```diff
@@ -0,0 +1,110 @@
+#include <QPropertyAnimation>
+#include <QTimer>
+
+#include <QDebug>
```
Is this still needed?

JamesMBallard commented on widgets/Esri/ArcGISRuntime/Toolkit/Flash.cpp:
```diff
@@ -0,0 +1,110 @@
+
+void Flash::setTargetColor(QColor color)
+{
+  m_color = color;
```
pass by const-ref or std::move.

JamesMBallard commented on widgets/Esri/ArcGISRuntime/Toolkit/Flash.cpp:
```diff
@@ -0,0 +1,110 @@
+  if (m_point == point)
+    return;
+
+  m_point = point;
```
pass by const-ref or std::move.

JamesMBallard commented on widgets/Esri/ArcGISRuntime/Toolkit/Flash.cpp:
```diff
@@ -0,0 +1,110 @@
+  auto animation = new QPropertyAnimation(this, "alpha", this);
+  animation->setKeyValues(QVariantAnimation::KeyValues { {0, 0}, {0.5, 255}, {1.0, 0} });
+  animation->setDuration(duration);
+  connect(animation, &QPropertyAnimation::finished, this, &QObject::deleteLater);
```
This is a really cool self-destruct mechanism 👍

JamesMBallard commented on widgets/Esri/ArcGISRuntime/Toolkit/Flash.h:
```diff
@@ -0,0 +1,129 @@
+
+  /*!
+   * \brief Set the radius of the circle of this flash.
+   * \param radius  Size of radius.
```
```suggestion
   * \param radius Size of radius.
```

JamesMBallard commented on widgets/Esri/ArcGISRuntime/Toolkit/Flash.h:
```diff
@@ -0,0 +1,129 @@
+   * 
+   * \param duration Lifetime of animation in ms.
+   */
+  void play(int duration);
```
since this is a slot, how about `onPlay` or `onPlayed`?

anmacdonald replied:
I think the `onSomething`  syntax tends to be more in the QML world. I don't know of any widgets that follow this style.

JamesMBallard commented on widgets/Esri/ArcGISRuntime/Toolkit/Flash.h:
```diff
@@ -0,0 +1,129 @@
+  void setAlpha(int alpha);
+
+  /*! \brief Gets the current alpha (for animation) */
+  int alpha() const;
```
optional add trailing `_` for private methods.
```suggestion
  int alpha_() const;
```

JamesMBallard commented on widgets/Esri/ArcGISRuntime/Toolkit/NorthArrow.h:
```diff
@@ -0,0 +1,92 @@
+ * compass heading matching the current rotation of the MapView, or Camera
+ * heading of the SceneView. 
+ * Double clicking on the NorthArrow triggers the heading of the connected
+ * GeoView to be orientainted to 0.
```
```suggestion
 * GeoView to be orientated to 0.
```

lsmallwood replied:
```suggestion
 * GeoView to be orientainted to 0 (North).
```

JamesMBallard commented on cpp/Esri/ArcGISRuntime/Toolkit/CoordinateConversionController.h:
```diff
@@ -0,0 +1,311 @@
+
+private:
+  Point m_currentPoint;
+  TaskWatcher m_screenToLocationProcess;
```
Recommendation:
```suggestion
  TaskWatcher m_screenToLocationTask;
```

lsmallwood commented on import/+cpp_api/CoordinateConversionController.qml:
```diff
@@ -0,0 +1,17 @@
+ *  limitations under the License.
+ ******************************************************************************/
+import Esri.ArcGISRuntime.Toolkit.CppInternal 1.0
+CoordinateConversionControllerCPP { }
```
I think it's worth a comment explaining what these do with the file selector

lsmallwood commented on import/CoordinateConversion.qml:
```diff
@@ -0,0 +1,327 @@
+                            return;
+
+                        var screenPos = coordinateConversionWindow.controller.screenCoordinate();
+                        if (screenPos === null || (screenPos.x === -1.0 && screenPos.y === -1.0))
```
i think it can if the point in question is off the screen

lsmallwood commented on import/CoordinateConversionController.qml:
```diff
@@ -0,0 +1,277 @@
+    property double zoomToDistance: 1500.0
+
+    /*!
+     * \brief Whether this controller is actively listening for mouse
```
maybe good to say these are mouse/touch events

lsmallwood commented on import/CoordinateConversionController.qml:
```diff
@@ -0,0 +1,277 @@
+
+    /*!
+    * \brief Updates the GeoView's camera to point to the current point's
+    * location on the map.
```
map or scene

lsmallwood commented on import/CoordinateConversionController.qml:
```diff
@@ -0,0 +1,277 @@
+        if (geoView === null)
+            return;
+
+        if (geoView.currentViewpointCamera !== undefined) {
```
undefined if the geoView is a map I'm guessing

anmacdonald replied:
Yeah, as Luke says, this is testing for method existence before calling.

lsmallwood commented on import/CoordinateConversionController.qml:
```diff
@@ -0,0 +1,277 @@
+    * So this forces _that_ controller to emit the currentPointChanged
+    * signal again: which force all downstream consumers to update.
+    * 
+    * If a C++ point were consumable in QML then we would just call the point
```
so this can all go away with api registration of Point?

anmacdonald replied:
It should do!

lsmallwood commented on import/NorthArrow.qml:
```diff
@@ -0,0 +1,78 @@
+        anchors.fill: parent
+        source: "images/compass.png"
+        fillMode: Image.PreserveAspectFit
+        visible: autoHide && (controller.heading < 1e-05 || controller.heading === 360) ? false : true
```
just a note that in some scenarios the heading can wrap around - so you need to handle 0, 360, 720 etc.. So you may want to mod heading divided by 360 and check the remainder or something

lsmallwood commented on import/NorthArrow.qml:
```diff
@@ -0,0 +1,78 @@
+        anchors.fill: parent
+        onDoubleClicked: {
+            if (geoView)
+                controller.setHeading(0);
```
feels like we should have a comment here to say "rotate back to north" or something

lsmallwood commented on register/Esri/ArcGISRuntime/Toolkit/register.cpp:
```diff
@@ -0,0 +1,93 @@
+namespace Toolkit
+{
+
+const char* const Namespace= "Esri.ArcGISRuntime.Toolkit.CppInternal";
```
Normally our convention is lowerCase first letter for variable names.

lsmallwood commented on register/Esri/ArcGISRuntime/Toolkit/register.cpp:
```diff
@@ -0,0 +1,93 @@
+
+namespace
+{
+
```
blank line

lsmallwood commented on register/Esri/ArcGISRuntime/Toolkit/register.cpp:
```diff
@@ -0,0 +1,93 @@
+
+/*!
+ * \internal
+ * \brief Convenience function for registration. Registers the C++ type Foo as
```
not really convenience - you _have_ to do it this way

lsmallwood commented on widgets/Esri/ArcGISRuntime/Toolkit/CoordinateConversion.cpp:
```diff
@@ -0,0 +1,159 @@
+#include <QItemDelegate>
+#include <QMenu>
+
+// ArcGISRuntime headers
```
normally our include order would put the lowest level libs lower done in the order. I think it should be:

- app files if appropriate
- toolkit files
- runtime api files
- qt
- stl

lsmallwood commented on widgets/Esri/ArcGISRuntime/Toolkit/CoordinateConversion.cpp:
```diff
@@ -0,0 +1,159 @@
+  QFrame(parent),
+  m_controller(new CoordinateConversionController(this)),
+  m_flash(nullptr),
+  m_ui(new Ui::CoordinateConversion)
```
this allows us to fwd declare the UI type in the header - i'd say this is a better pattern for widgets UI

anmacdonald replied:
Probably can keep forward declared for now.

lsmallwood commented on widgets/Esri/ArcGISRuntime/Toolkit/CoordinateConversion.cpp:
```diff
@@ -0,0 +1,159 @@
+    addMenu->addAction(format->name(), m_controller,
+    [this, menu, f(QPointer<CoordinateConversionOption>(format))]()
+    {
+      if (f) m_controller->addNewCoordinateResultForOption(f.data());
```
add brackets and new line

lsmallwood commented on widgets/Esri/ArcGISRuntime/Toolkit/CoordinateConversion.cpp:
```diff
@@ -0,0 +1,159 @@
+    return;
+
+  const auto point = m_controller->screenCoordinate();
+  if (point.isNull() || isnan(point.x()) || isnan(point.y()))
```
probs need `cmath` include for this

lsmallwood commented on widgets/Esri/ArcGISRuntime/Toolkit/Flash.cpp:
```diff
@@ -0,0 +1,110 @@
+{
+  if (m_color.alpha() == alpha)
+    return;
+  m_color.setAlpha(alpha);
```
add line after the `return`

lsmallwood commented on widgets/Esri/ArcGISRuntime/Toolkit/NorthArrow.cpp:
```diff
@@ -0,0 +1,90 @@
+#include "NorthArrow.h"
+
+// Qt headers
+#include <QMouseEvent>
```
reorder include blocks

lsmallwood commented on widgets/Esri/ArcGISRuntime/Toolkit/NorthArrow.cpp:
```diff
@@ -0,0 +1,90 @@
+
+void NorthArrow::mouseDoubleClickEvent(QMouseEvent* event)
+{
+  m_controller->setHeading(0);
```
rotate back to north comment

anmacdonald commented on README.md:
````diff
@@ -1,152 +1,168 @@
 
-```
-include(path/to/toolkit/toolkitwidgets.pri)
+```pro
````
Ahh, very handy link! I just made a guess.

anmacdonald commented on cpp/Esri/ArcGISRuntime/Toolkit/CoordinateConversionOption.cpp:
```diff
@@ -0,0 +1,249 @@
+    return;
+
+  m_garsConvesrionMode = conversionMode;
+  emit garsConversionModeChanged();
```
I am on the fence here! I am not sure myself? Do you think it'll be worth it in the long term?

JamesMBallard replied:
Let's keep what you have for now. It will be easier to read and debug for us and users. If we need to write a lot more of these, we can consider a helper of some kind.

anmacdonald commented on import/CoordinateConversion.qml:
```diff
@@ -0,0 +1,327 @@
+                        var itemPos = geoView.mapToItem(geoView, screenPos.x, screenPos.y);
+                        checked = true;
+                        var flashImage = internal.flashImageFactory.createObject(geoView, { "x": itemPos.x, "y": itemPos.y, "color": highlightColor });
+                        flashImage.finished.connect(function() { flashCoordinateButton.checked = false; });
```
Yep you got it! Is a custom signal fired on end of animation.

anmacdonald commented on import/CoordinateConversion.qml:
```diff
@@ -0,0 +1,327 @@
+                    Layout.maximumWidth: Layout.maximumHeight
+                    padding: 0
+                }
+        }
```
Good eye, yes it looks like there was!

anmacdonald commented on import/CoordinateConversion.qml:
```diff
@@ -0,0 +1,327 @@
+    QtObject {
+        id: internal
+        property Component flashImageFactory: Qt.createComponent("FlashImage.qml");
+    }
```
It's not needed necessarilly, but I want to make it inaccessabile to public consumers.

JamesMBallard replied:
That makes sense.

anmacdonald commented on import/CoordinateConversionController.qml:
```diff
@@ -0,0 +1,277 @@
+    * \brief Known list of available coordinate conversion formats which can be
+    * consumed to generate different textual representations of the same point.
+    */
+    property list<CoordinateConversionOption> formats: [
```
Hmm, this doesn't seem to be working. I get a syntax error when I do this. Is there a special way to initialize a `var` type with a `list<T>`?

JamesMBallard replied:
It should be just `property var formats: [ stuff here... ]`

anmacdonald replied:
It looks as if my QML code isn't happy if I try to assign to the list using the QML declarative syntax, unless I explicitly let QML know I'm assinging to a `list<T>`? Not sure how to get around this one. I think we might be stuck with `list<T>` unless I am missing something.

JamesMBallard replied:
Ok, no problem then. Keep what you have.

anmacdonald commented on import/CoordinateConversionController.qml:
```diff
@@ -0,0 +1,277 @@
+            return Qt.point(-1, -1);
+
+        let point = geoView.locationToScreen(internal.currentPoint);
+        if (point && geoView.geoViewType === Enums.GeoViewTypeSceneView)
```
I think it might in QML?

anmacdonald commented on import/CoordinateConversionController.qml:
```diff
@@ -0,0 +1,277 @@
+            // Point is  a string, so convert to a point using the option param.
+            if (option !== undefined) {
+                let sp = geoView ?  geoView.spatialReference : undefined;
+                internal.currentPoint = option.pointFromString(point, sp);
```
I tried to copy the C++ logic which passed a default-constructed Spatial reference into the `CoordinateFormatter`. Is there any guarnatee that the QML has parity in this way? I.E. that a null value is equivalent to an invalid default-constructed value type?

JamesMBallard replied:
>Is there any guarnatee that the QML has parity in this way? I.E. that a null value is equivalent to an invalid default-constructed value type?

No, cannot guarantee that. There is a common "empty"/"isEmpty" property that could either mean it is default-constructed, or it is a valid SpatialReference from core that happens to be "empty". We cannot differentiate between the two in the pubic C++ Api from user code, so there is no way to equate `(Qml)SpatialReference == null` implies C++ SpatialReference object is default-constructed.

anmacdonald commented on import/CoordinateConversionController.qml:
```diff
@@ -0,0 +1,277 @@
+    * returned by coordinateResults.
+    */
+    function removeCoordinateResultAtIndex(index) {
+        internal.results.splice(index, 1);
```
If splice behaves as documented by [MDN](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/splice), it shouldn't need validated.

anmacdonald commented on import/CoordinateConversionController.qml:
```diff
@@ -0,0 +1,277 @@
+            });
+            geoView.setViewpointCameraAndSeconds(newCamera, 1.0);
+        }
+        else if (geoView.mapRotation !== undefined) {
```
Again, checking method existence.

anmacdonald commented on register/Esri/ArcGISRuntime/Toolkit/register.cpp:
```diff
@@ -0,0 +1,93 @@
+#include "Esri/ArcGISRuntime/Toolkit/CoordinateConversionController.h"
+#include "Esri/ArcGISRuntime/Toolkit/CoordinateConversionOption.h"
+#include "Esri/ArcGISRuntime/Toolkit/CoordinateConversionResult.h"
+#include "Esri/ArcGISRuntime/Toolkit/NorthArrowController.h"
```
This is a consequence of building ourselves directly into the user's project.

Anything we add to the `INCLUDEPATH` is also available to the user and to ourselves.

The choice becomes: do we want to impose on the user that toolkit imports are namespace qualified or not? Should a user write `#include "Esri/ArcGISRuntime/Toolkit/NorthArrowController.h"`, or can they just write `#include "NorthArrowController.h"`?

JamesMBallard replied:
>The choice becomes: do we want to impose on the user that toolkit imports are namespace qualified or not? Should a user write #include "Esri/ArcGISRuntime/Toolkit/NorthArrowController.h", or can they just write #include "NorthArrowController.h"?

Those are two different questions to me. The namespace and the folder structure do not need to match. The users needs `NorthArrowController.h`, so I would prefer if they can just write `#include "NorthArrowController.h"` and not be concerned with the folder structure.

anmacdonald commented on toolkitqml.pri:
```diff
@@ -1,5 +1,18 @@
+RESOURCES += $$PWD/images/esri_arcgisruntime_toolkit_images.qrc \
+             $$PWD/import/esri_arcgisruntime_toolkit_qml.qrc
 
+# For Qt Creator only
```
This is now moot now everything is in qr files. Removing.

anmacdonald commented on widgets/Esri/ArcGISRuntime/Toolkit/CoordinateConversion.cpp:
```diff
@@ -0,0 +1,159 @@
+  if (point.isNull() || isnan(point.x()) || isnan(point.y()))
+    return;
+
+  m_flash = new Flash();
```
As you probably guessed now, flash self-deletes.

`Flash` is reparented to the scene when the widget is added, so a unique_ptr would not be appropriate as if is now in the child-parent hierarchy.

It's all a bit mungy, I know.

anmacdonald commented on widgets/Esri/ArcGISRuntime/Toolkit/CoordinateConversion.cpp:
```diff
@@ -0,0 +1,159 @@
+#include "Esri/ArcGISRuntime/Toolkit/CoordinateConversionController.h"
+#include "Esri/ArcGISRuntime/Toolkit/CoordinateConversionResult.h"
+#include "Esri/ArcGISRuntime/Toolkit/GenericListModel.h"
+#include "Esri/ArcGISRuntime/Toolkit/GenericTableProxyModel.h"
```
See comment about INCLUDEPATHS.

anmacdonald commented on widgets/Esri/ArcGISRuntime/Toolkit/CoordinateEditDelegate.cpp:
```diff
@@ -0,0 +1,84 @@
+#include "Esri/ArcGISRuntime/Toolkit/CoordinateConversionResult.h"
+#include "Esri/ArcGISRuntime/Toolkit/CoordinateConversionController.h"
+#include "Esri/ArcGISRuntime/Toolkit/GenericListModel.h"
+#include "Esri/ArcGISRuntime/Toolkit/GenericTableProxyModel.h"
```
See comment about INCLUDEPATHS.

anmacdonald commented:
I have addressed the PR feedback on stylistic changes.
Structure changes are as follows:

* INCLUDEPATH is now qualified. The user's include now looks like `#include "NorthArrow.h"` if they so want.
* Point is registered as a metatype and workaround code is now removed.
* GenericListModel/Flash etc now are within an `Internal` folder to make their internal nature more clear.

lsmallwood commented on register/Esri/ArcGISRuntime/Toolkit/register.h:
```diff
@@ -45,6 +45,14 @@ namespace Toolkit
  * Internally, this function add a new `cpp_api` selector to your QmlEngine's
  * file selector. This is the mechanism utilized to override, say,
  * NorthArrowController.qml with the NorthArrowController provided by C++.
+ *
```
remove extra blank line

JamesMBallard commented:
This is looking really good now. The addition of the internal folder should really help users know what they should be focusing on, and specifically what _not_ to be concerned with.

anmacdonald commented on cpp/Esri/ArcGISRuntime/Toolkit/CoordinateConversionController.h:
```diff
@@ -18,7 +18,7 @@
 // Toolkit headers
 #include "CoordinateConversionOption.h"
-#include "GenericListModel.h"
+#include "Internal/GenericListModel.h"
```
In this case - the `Q_PROPERTY` declaration needs to recognize the the inheritance hierarchy, as in QML the `GenericListModel` is only exposed as an `AbstractListModel`.
